<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>SeqWins Vignette</title>

<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>
<style type="text/css">
a.anchor-section {margin-left: 10px; visibility: hidden; color: inherit;}
a.anchor-section::before {content: '#';}
.hasAnchor:hover a.anchor-section {visibility: visible;}
</style>
<script>// Anchor sections v1.0 written by Atsushi Yasumoto on Oct 3rd, 2020.
document.addEventListener('DOMContentLoaded', function() {
  // Do nothing if AnchorJS is used
  if (typeof window.anchors === 'object' && anchors.hasOwnProperty('hasAnchorJSLink')) {
    return;
  }

  const h = document.querySelectorAll('h1, h2, h3, h4, h5, h6');

  // Do nothing if sections are already anchored
  if (Array.from(h).some(x => x.classList.contains('hasAnchor'))) {
    return null;
  }

  // Use section id when pandoc runs with --section-divs
  const section_id = function(x) {
    return ((x.classList.contains('section') || (x.tagName === 'SECTION'))
            ? x.id : '');
  };

  // Add anchors
  h.forEach(function(x) {
    const id = x.id || section_id(x.parentElement);
    if (id === '') {
      return null;
    }
    let anchor = document.createElement('a');
    anchor.href = '#' + id;
    anchor.classList = ['anchor-section'];
    x.classList.add('hasAnchor');
    x.appendChild(anchor);
  });
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">SeqWins Vignette</h1>



<div id="introduction" class="section level3">
<h3>Introduction</h3>
<p>SeqWins (fastq <strong>Seq</strong>uence analysis on <strong>Win</strong>dows system) can achieve on pure Windows system flexible base-level (ATCG) quality control and convenient high-level whole-process analysis of fastq data– spanning quality control report, trimming bases and reads accordingly, alignment and feature count. Till now, words still prevail that Fastq data cannot be elegantly processed in Windows. However, the fundamental low-level R package <a href="https://bioconductor.org/packages/ShortRead/">ShortRead</a> and <a href="https://bioconductor.org/packages/Rsubread/">Rsubread</a> has been available for long, based on which SeqWins is build up as complementary to other useful R packages, like <a href="https://bioconductor.org/packages/QuasR/">QuasR</a> etc. Hope you’d find SeqWins useful and saving you from struggling for days to set up system environment and try various tools seemingly compatible with you daily Windows work station like I did when carrying out relevant tasks.</p>
<hr />
</div>
<div id="prepare-genomic-fastq-sequence-and-the-corresonding-gtf-annotation" class="section level3">
<h3>Prepare genomic Fastq sequence and the corresonding GTF annotation</h3>
<p>These gemonic data of the same species as with your fastq data are required to build index when aligning reads and get feature counts, and can be download from NCBI assembly website, for example <a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.26/">the hg38 files</a> by clicking the top right “<strong>Download Assembly</strong>” button and selecting “<strong>Genomic FASTA (.fna)</strong>” and “<strong>Genomic GTF (.gtf)</strong>” sequentially on the resultant drop-down options.</p>
<p>If necessary untar the downloaded files as blow</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">untar</span>(<span class="st">&quot;full/path/genome_assemblies_genome_fasta.tar&quot;</span>,<span class="dt">exdir =</span> <span class="st">&quot;.&quot;</span>)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw">untar</span>(<span class="st">&quot;full/path/genome_assemblies_genome_gtf.tar&quot;</span>,<span class="dt">exdir =</span> <span class="st">&quot;.&quot;</span>)</span></code></pre></div>
</div>
<div id="qc-report" class="section level3">
<h3>QC report</h3>
<p>set working directory, it will be where the QC result folder and index file are located and <code>./bam</code> will be where the bam files is located</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw">setwd</span>(<span class="st">&quot;/your/working/directory&quot;</span>)</span></code></pre></div>
<p>and get QC report</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># call ShortRead::qa to generate QC reports of all fastq.gz files</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>qa &lt;-<span class="st"> </span><span class="kw">qa</span>(<span class="st">&quot;full/path/FastqFolder&quot;</span>, <span class="st">&quot;fastq.gz&quot;</span>)</span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="kw">browseURL</span>(<span class="kw">report</span>(qa))</span></code></pre></div>
<p>the above will generate report of the fastq files that can be retrieved by</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw">list.files</span>(<span class="dt">path =</span> <span class="st">&quot;full/path/FastqFolder&quot;</span>,<span class="dt">pattern =</span> <span class="st">&quot;fastq.gz&quot;</span>)</span></code></pre></div>
</div>
<div id="trim-basesfilter-readsalign-and-count-feature" class="section level3">
<h3>Trim bases,filter reads,align and count feature</h3>
<p>According to the QC report, customize the parameters used for trimming bases, filtering reads. Although most can be left default,please specify file paths totally decided by yourself; the “subReadThreads”,“shortreadRAM” decided by your computer; and the sequence tech type decided by your project,hopefully not by money.</p>
<div id="set-aligntype" class="section level4">
<h4><em>set alignType</em></h4>
<p>Specifically for RNAseq leave <code>alignType</code> as default (simply don’t mention it). By the way, unit of “shortreadRAM” is byte,don’t panic.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="kw">seqW</span>(<span class="dt">fileList1=</span>fl,<span class="dt">subReadThreads=</span>3L,<span class="dt">shortreadRAM=</span><span class="fl">1e8</span>,</span>
<span id="cb5-2"><a href="#cb5-2"></a>                <span class="dt">genomeRefFile=</span><span class="st">&quot;./GCF_000001405.26_GRCh38_genomic.fna.gz&quot;</span>,</span>
<span id="cb5-3"><a href="#cb5-3"></a>                <span class="dt">genomeAnnotFile=</span><span class="st">&quot;./GCF_000001405.39_GRCh38.p13_genomic.gtf.gz&quot;</span>)</span></code></pre></div>
<p>For DNAseq, longDNAseq, RNAseq aim at exon-exon junction and microRNAseq, set <code>alignType</code> as ‘<code>dna</code>’,‘<code>dnaLong</code>’,‘<code>rnaExon_Exon</code>’and’<code>microRNA</code>’ respectively–</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="kw">seqW</span>(<span class="dt">fileList1=</span>fl,<span class="dt">alignType =</span><span class="st">&#39;dna&#39;</span>)<span class="co">#DNAseq</span></span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="kw">seqW</span>(<span class="dt">fileList1=</span>fl,<span class="dt">alignType =</span><span class="st">&#39;dnaLong&#39;</span>)<span class="co">#longDNAseq</span></span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="kw">seqW</span>(<span class="dt">fileList1=</span>fl,<span class="dt">alignType =</span><span class="st">&#39;rnaExon_Exon&#39;</span>)<span class="co">#RNAseq aim at exon-exon junction</span></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="kw">seqW</span>(<span class="dt">fileList1=</span>fl,<span class="dt">alignType =</span><span class="st">&#39;microRNA&#39;</span>)<span class="co">#microRNAseq</span></span></code></pre></div>
</div>
<div id="specify-file-path" class="section level4">
<h4><em>specify file path</em></h4>
<p>If your fastq were generated from single-end sequencing (old-fashioned you are!), only specify <code>fileList1</code> as blow</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>fl&lt;-<span class="kw">list.files</span>(<span class="dt">path =</span> <span class="st">&quot;full/path/FastqFolder&quot;</span>,<span class="dt">pattern =</span> <span class="st">&quot;fastq.gz&quot;</span>)</span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="kw">seqW</span>(<span class="dt">fileList1=</span>fl,<span class="dt">genomeRefFile=</span><span class="st">&quot;./GCF_000001405.26_GRCh38_genomic.fna.gz&quot;</span>,</span>
<span id="cb7-3"><a href="#cb7-3"></a>                <span class="dt">genomeAnnotFile=</span><span class="st">&quot;./GCF_000001405.39_GRCh38.p13_genomic.gtf.gz&quot;</span>)<span class="co">#RNAseq</span></span></code></pre></div>
<p>If paired-end,specify both <code>fileList1</code> i.e. the “read 1” files and <code>fileList2</code> i.e. “read 2” files</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>fl_<span class="dv">1</span>&lt;-<span class="kw">list.files</span>(<span class="dt">path =</span> <span class="st">&quot;full/path/FastqFolder&quot;</span>,<span class="dt">pattern =</span> <span class="st">&quot;^.*_1</span><span class="ch">\\</span><span class="st">.fastq</span><span class="ch">\\</span><span class="st">.gz$&quot;</span>)</span>
<span id="cb8-2"><a href="#cb8-2"></a>fl_<span class="dv">2</span>&lt;-<span class="kw">list.files</span>(<span class="dt">path =</span> <span class="st">&quot;full/path/FastqFolder&quot;</span>,<span class="dt">pattern =</span> <span class="st">&quot;^.*_2</span><span class="ch">\\</span><span class="st">.fastq</span><span class="ch">\\</span><span class="st">.gz$&quot;</span>)</span>
<span id="cb8-3"><a href="#cb8-3"></a></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="kw">seqW</span>(<span class="dt">fileList1=</span>fl_<span class="dv">1</span>,<span class="dt">fileList1=</span>fl_<span class="dv">2</span>,<span class="dt">genomeRefFile=</span><span class="st">&quot;./GCF_000001405.26_GRCh38_genomic.fna.gz&quot;</span>,</span>
<span id="cb8-5"><a href="#cb8-5"></a>                <span class="dt">genomeAnnotFile=</span><span class="st">&quot;./GCF_000001405.39_GRCh38.p13_genomic.gtf.gz&quot;</span>)<span class="co">#RNAseq</span></span></code></pre></div>
<p>Please don’t forget to specify <code>genomeRefFile</code> and <code>genomeAnnotFile</code> i.e. the path of “<strong>Genomic FASTA (.fna)</strong>” and “<strong>Genomic GTF (.gtf)</strong>” as any of the above examples</p>
</div>
<div id="specify-trim-and-filter" class="section level4">
<h4><em>specify trim and filter</em></h4>
<p>Below shows a complete list of trim and filter parameters. Probably most can be left as default except the adaptor sequence specific to your sequence platform</p>
<ol style="list-style-type: decimal">
<li><code>endTrimThrs</code></li>
</ol>
<p>hred score threshold of the end base below which the end base will be trimmed,default “?”</p>
<ol start="2" style="list-style-type: decimal">
<li><code>endTrimThrsend</code></li>
</ol>
<p>mean Phred score threshold of five bases in ends,below which the five bases will be trimmed,default “4”</p>
<ol start="3" style="list-style-type: decimal">
<li><code>adpter1Seq</code>,<code>adpter2Seq</code></li>
</ol>
<p>adapter sequence to be trimmed from end and also inner part of the read</p>
<ul>
<li><p>default <code>adpter1Seq</code> “AGATCGGAAGAGCACACGTCTGAACTCCAGTCA” i.e. illumina HT4000 adapter for read 1</p></li>
<li><p>default <code>adpter2Seq</code> “AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT” i.e. illumina HT4000 adapter for read 2</p></li>
</ul>
<ol start="4" style="list-style-type: decimal">
<li><code>with.Lindels</code>,<code>with.Rindels</code></li>
</ol>
<p>True if your reads contain indels on left end (5’,corresponding to <code>with.Lindels</code>) or right end (3’,corresponding to <code>with.Rindels</code>), default False</p>
<ol start="5" style="list-style-type: decimal">
<li><code>widthThrs</code></li>
</ol>
<p>width threshold of reads to be filtered out,default 14L</p>
<ol start="6" style="list-style-type: decimal">
<li><code>cmplxThrs</code></li>
</ol>
<p>complexity threshold of reads to be filtered out,default 0.5 i.e. half of mean complexity of human genome</p>
<ol start="7" style="list-style-type: decimal">
<li><code>innerNThrs</code></li>
</ol>
<p>number of N inside the read below which the read will be removed,default 2L</p>
<ul>
<li><p><code>innerS</code></p>
<p>which base should be considered the start of inside part of the read after the above trimming,default 4L</p></li>
<li><p><code>innerE</code></p>
<p>which base should be considered the end of inside part of the read after the above trimming. Negative integer X for width(read)-abs(X) default -4L; positive integer X for Xth base</p></li>
</ul>
</div>
<div id="summary-of-trimming-and-filtration" class="section level4">
<h4><em>summary of trimming and filtration</em></h4>
<p>A list containing dataframes of which each reports the result of trimming and filtration will be returned by the <code>seqw</code> function. You can assign the list to an object for further observation.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>reportList&lt;-<span class="kw">seqW</span>(<span class="dt">fileList1=</span>fl_<span class="dv">1</span>,<span class="dt">fileList1=</span>fl_<span class="dv">2</span>,<span class="dt">genomeRefFile=</span><span class="st">&quot;./GCF_000001405.26_GRCh38_genomic.fna.gz&quot;</span>,</span>
<span id="cb9-2"><a href="#cb9-2"></a>                <span class="dt">genomeAnnotFile=</span><span class="st">&quot;./GCF_000001405.39_GRCh38.p13_genomic.gtf.gz&quot;</span>)<span class="co">#RNAseq</span></span></code></pre></div>
</div>
</div>
<div id="even-more-flexibility-is-supported" class="section level3">
<h3>Even more flexibility is supported!</h3>
<p>Hope the package can realize easy but deep appreciation of your fastq data, and you probably have found every parameters you wanna touch through above introduction. But there is even more! <code>...</code> arguments enable <code>seqW</code> function to access all the arguments in the functions wrapped inside it, including</p>
<ol style="list-style-type: decimal">
<li><code>Rsubread::align</code>, called when <code>alignType=&quot;dna&quot;</code> , <code>&quot;rna&quot;</code> ,<code>&quot;microRNA&quot;</code></li>
<li><code>Rsubread::sublong</code>, called when <code>alignType=&quot;longDNA&quot;</code></li>
<li><code>Rsubread::subjunc</code>, called when <code>alignType=&quot;rnaExon_Exon&quot;</code></li>
<li><code>Rsubread::featureCounts</code>, called every time <code>seqW</code> runs ( thus simply add their parameters (see what they have by <code>?Rsubread::featureCounts</code> or <a href="https://bioconductor.org/packages/release/bioc/vignettes/Rsubread/inst/doc/SubreadUsersGuide.pdf">Rsubread vignettes</a>) when calling seqW, e.g.</li>
</ol>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="co"># minFragLength by default is 50L in Rsubread, but you can modify it easily as blow</span></span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="kw">seqW</span>(<span class="dt">minFragLength=</span>40L,<span class="dt">fileList1=</span>fl,<span class="dt">genomeRefFile=</span><span class="st">&quot;./GCF_000001405.26_GRCh38_genomic.fna.gz&quot;</span>,</span>
<span id="cb10-3"><a href="#cb10-3"></a>                <span class="dt">genomeAnnotFile=</span><span class="st">&quot;./GCF_000001405.39_GRCh38.p13_genomic.gtf.gz&quot;</span>)<span class="co">#RNAseq</span></span></code></pre></div>
<p>You might even wanna add/alter the trimming and filtration pipeline; No problem! I have exported every units wrapped under the top-level function, including</p>
<ol style="list-style-type: decimal">
<li><code>trimEnds</code></li>
<li><code>trimTailw</code></li>
<li><code>trimAdapter</code></li>
<li><code>filterWidth</code></li>
<li><code>filterLowQuality</code></li>
<li><code>nFilter</code></li>
<li><code>filterLowComplexity</code></li>
<li><code>filterInnerN</code></li>
</ol>
<p>simply <code>?trimEnds</code> for example, to see their function and how to use it as bricks to make your own trimming and filter flow. And then use aforementioned <code>Rsubread</code> functions to complete alignment and feature count.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
